### ISSUE TITLE
[NullnessChecker/jdk stubs] allow @Nullable for `Objects.requireNonNull`

### ISSUE THREAD

**Issue by xenoterracide** [typetools/checker-framework#4031]
given

```java
Objects.requireNonNull(null);
```
ok that's a silly example, but literally the return of anything that is marked with `@Nullable`. I would argue this should be allowed as is, however I would be ok if the message were required in order to insure there is a reasonable message.

```java
Objects.requireNonNull(null, "I always throw an exception);
```
where the above would still provide an error due to not null. Not certain if a better error message could be provided to say use the version with a message, or perhaps if further documentation could somehow be provided inline that shows up in an IDE.

**Comment by mernst** [comment]
To repeat two comments from a previous conversation:
 * [The manual](https://checkerframework.org/manual/#annotate-normal-behavior-always-crash) explains that `@NonNull` is the correct annotation.
 * If you can suggest how the manual can be improved, please do so.  (Maybe the current text is too wordy or indirect.)

If you don't care that your program crashes due to `Objects.requireNonNull`, but you do care if your program crashes due to any other method that is wrongly passed `null`, you can use the [`-AskipUses`](https://checkerframework.org/manual/#askipuses) command-line option to suppress warnings about uses of the `Objects` class.

**Comment by xenoterracide** [comment]
> ou can use the -AskipUses command-line option to suppress warnings about uses of the Objects class.

that would skip them all, and I don't know that's what I want.

> The manual explains that @NonNull is the correct annotation.

I still disagree with this, for this method explicitly. Which is why I made a separate ticket.

**Comment by mernst** [comment]
> I still disagree with this, for this method explicitly. Which is why I made a separate ticket.

Can you please explain why you disagree?  Just saying "I disagree" does not lead to enlightenment on either side. :-)

**Comment by rkraneis** [comment]
I guess this goes in the same direction as the `@Present  this` annotations on `Optional#getOrThrow` and similar requirements in `Opts`, which I already commented on in the past. I actually am in the same camp as @xenoterracide (which means we do want to treat these methods as ways to explicitly stop execution of the program), but I understand checkers philosophy to make that a bit harder and just forbid unsafe use of framework methods.

**Comment by RamAnvesh** [comment]
From the manual, while I understand that `a.get()` is linguistically similar to `requireNonNull(a).get()`, semantically they **_may be seen_** differently. One is saying: I am assuming `a` cannot be null. Other is saying: I want `a` to be non-null, and if it is not, I want the program to crash. 
This is very useful. For example, in cases where I just added a key to a map/or a DB, and I know if it is missing now, I want the program to crash. What is the correct way to convey this intent in the program? 

In other libraries like guava, `Preconditions.checkNotNull` accepts a non-null param. Which makes sense because this is a validation. Clients who are passing arguments to this method should have presumably marked the passed variable as `@NonNull` in the first place.

But in the same guava, `Verify.verifyNotNull` accepts a `@Nullable` which also makes sense because I want to be able to pass nullable references to it so that it can crash as per my design, with my full awareness.

If these both are correct, then the statement from the manual `Method m2 behaves just like m1 in that it throws NullPointerException if a client passes null. Therefore, a checker should issue the same warning if a client might pass null.`
Is not exactly correct. Just because `requireNonNull(a).get()` linguistically behaves exactly like `a.get()` doesn't mean they are semantically same.
So then the discussion becomes whether `requireNonNull` is in the same family as `Preconditions.checkNotNull` or `Verify.verifyNonNull`. From the documentation of `requireNonNull` it seems the former. So I agree that the current state is affairs is acceptable.

That being said:

1. Can you please share your opinion on `Verify.verifyNotNull`? Is their use of `@Nullable` correct according to you?
2. If you agree that `Verify.verifyNotNull` is correct in its usage of Nullable, and you agree with the above arguments (that `requireNonNull`'s documentation says that it is used for validation, hence it should **_not_** accept nullable references), can you please update the manual focusing on the semantic similarity rather than the linguistic similarity?
3. Can you please update the manual with this difference (between validators and assertions/verifiers) and point to methods like `Verify.verifyNotNull` so that users can write such code without warnings:
```
String name = names.get("123");
//Name should not be null. If it null, explode!
System.out.println("Name length:" + verifyNotNull(name).length());
```

**Comment by mernst** [comment]
Thanks for your explanation.  I appreciate it.

I got confused right at the beginning:

> From the manual, while I understand that `a.get()` is linguistically similar to `requireNonNull(a).get()`, semantically they **_may be seen_** differently. One is saying: I am assuming `a` cannot be null, or more precisely, I have not given the nullness of `a` any thought. Other is saying: I want `a` to be non-null, and if it is not, I want the program to crash. 

`a.get()` and `requireNonNull(a).get()` are not merely "similar".  They have *identical* behavior.  Both of them mean:  if `a` is null, the program crashes.  The stack trace is even the same in both cases.

The purpose of `requireNonNull` is to obtain an exception earlier.  I see that in your personal usage, you overlay it with another meaning, namely "I have examined this code and believe it to be correct, so I do not wish tools to warn me about problems with it."  That's fine for you to do, but it is non-standard, so tools do not cater to this non-standard usage.  (As a matter of style, I would not want my code cluttered with `requireNonNull` at every single usage that I think is correct.  I would rather mark the parts that I am concerned might be buggy.  But, to each his own.)  You can customize the Checker Framework to treat it this way; see below.

> 1. Can you please share your opinion on `Verify.verifyNotNull`? Is their use of `@Nullable` correct according to you?

You are right that [`Verify.verifyNotNull` specifies a `@Nullable` formal parameter type](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Verify.html#verifyNotNull(T)).  This is an incorrect specification.  It probably results from two problems:
 * The Guava authors are on record saying their goal in writing nullness annotations is to "make FindBugs happy", not to make their specifications clear, consistent, or correct.
 * The FindBugs definition of `@Nullable` is [nonsense](https://checkerframework.org/manual/#findbugs-nullable).

The Nullness Checker uses a different, more principled definition of `@Nullable`.  [The Javadoc of `@Nullable`](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/Nullable.html) says

> `Nullable` is a type annotation that indicates that the value is not known to be non-null (see `NonNull`).

[The Javadoc of `@NonNull`](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/NonNull.html) says:

> If an expression's type is qualified by `@NonNull`, then the expression never evaluates to `null`. (Unless the program has a bug; annotations specify intended behavior.)

For `verifyNotNull`, the formal parameter never evaluates to `null` unless the program has a bug, so its formal parameter type should be `@NonNull`.  If you see a problem with the Checker Framework's definition of `@Nullable` and `@NonNull`, let us know.

> 3. Can you please update the manual with this difference (between validators and assertions/verifiers) and point to methods like `Verify.verifyNotNull` for users like me who are interested in writing such code without checker warnings:

Wishing to suppress warnings is a reasonable goal.  I support that goal!
I think our discussion would be better framed as a desire to suppress some Nullness Checker warnings, rather than a discussion of some semantic difference.  That is, the difference is programmer intent, not semantics.

If your goal is to suppress warnings, then the Checker Framework offers a number of mechanisms to do so.
For your use case, probably the easiest are using `-AskipUses` or writing a stub to override the library annotation.

The manual discusses [the difference between defensive programming and specification](https://checkerframework.org/manual/#defensive-programming).  https://github.com/typetools/checker-framework/pull/4267 adds an explicit mention of `Objects.requireNonNull`, `verifyNotNull`, and `checkNotNull`.  Hopefully that will make it more discoverable by users.  https://github.com/typetools/checker-framework/pull/4264 also explained how to override library annotations if you wish to use a different specification, though #4267 expands on that.

(As a minor point that is probably moot, I don't understand the distinction you are making between "validation" and "verification".  It does not conform to the [standard definition](https://en.wikipedia.org/wiki/Verification_and_validation) of these terms in computer science, so you might want to use different words or give definitions for what you mean when you say them.  I'm not trying to be pedantic, just to understand your points in a precise way.)

Thanks again for bringing up these points.  I hope this explanation helps.  If I have misconstrued you, please correct my misunderstanding.

**Comment by RamAnvesh** [comment]
Thanks a lot for your response. Things are a lot clearer to me now.

>  I don't understand the distinction you are making between "validation" and "verification"
You are correct - I might have misused the term verification. 
By validation I mean parameter validation, for example, checking that callers are passing non-null values in accordance with the method's javadoc specification (`Preconditions.checkNotNull` and `Objects.requireNonNull` fall in this category - such methods should not have `@Nullable` on the method param).
The word verification I took from the guava Verify class - which is incorrect. What I actually was referring to is to capture the programmer's intent: "I am able to prove that this reference is non-null. Please suppress all warnings". Having such a method allows me to write code like `notNull(notNull(a).getB()).getC()` - not ideal - but absolutely needed in some circumstances

> The manual discusses the difference between defensive programming and specification.
This [section](https://checkerframework.org/manual/#defensive-programming) clearly documents by concerns and requirements. And also mentions [NullnessUtil.castNonNull](https://checkerframework.org/api/org/checkerframework/checker/nullness/util/NullnessUtil.html#castNonNull-T-) which is exactly what I was looking for. I will be using this for the later use case from now on.

Thanks a lot for your time and detailed explanation!

**Comment by mernst** [comment]
I'm glad this is clearer.  Thank you for prompting us to improve the documentation.  Let us know if you have any other questions or suggestions.
